---
title: "Exp 005-CO610 karyotapR analysis"
author: "Joey Mays"
date: '2023-07-26'
output:
  pdf_document:
    toc: yes
  html_notebook:
    code_folding: show
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
    theme: flatly
    df_print: paged
  html_document:
    code_folding: show
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
    theme: flatly
    df_print: paged
editor_options:
  chunk_output_type: inline
  markdown:
    wrap: 72
---

# Setup

```{r message=FALSE}
library(karyotapR)
library(tidyverse)
library(ggplot2)
library(ggdist)
library(here)
```

```{r}
set.seed(20230726)
```

```{r}
here::i_am("analysis-notebooks/exp05-karyotapR-analysis.Rmd")
```

# Load Data

```{r}
tapestri.filename <- here("datasets", "tapestri-experiment05-panelv2.h5")
```

# Construct Object

```{r}
exp5 <- createTapestriExperiment(tapestri.filename, panel.id = "CO610")
```

# AF Clustering

## Embedding

Remove variants from gRNA probe

```{r}
keep.var <- which(rowData(altExp(exp5, "alleleFrequency"))$chr != "pLENTI-gRNA")
altExp(exp5) <- altExp(exp5)[keep.var, ]
```

```{r}
exp5 <- runPCA(exp5, "alleleFrequency", sd.min.threshold = 25)
PCAKneePlot(exp5)
```


```{r}
reducedDimPlot(exp5, dim.reduction = "pca")
```

```{r}
exp5 <- runUMAP(exp5, alt.exp = "alleleFrequency", use.pca.dims = T, pca.dims = 1:2)
reducedDimPlot(exp5, alt.exp = "alleleFrequency", dim.reduction = "umap")
```

## Clustering

```{r}
exp5 <- runClustering(exp5, alt.exp = "alleleFrequency", eps = 2)
reducedDimPlot(exp5, dim.reduction = "umap", group.label = "cluster")
```

Number of cells in each cluster

```{r}
fct_count(colData(exp5)$cluster)
```

Filter

```{r}
exp5.subset <- exp5[,colData(exp5)$cluster %in% 1:2]
```

```{r}
reducedDimPlot(exp5.subset, dim.reduction = "umap", group.label = "cluster")
```


## Probes

```{r}
summary(rowData(exp5.subset)$median.reads > 0)
rowData(exp5.subset)[rowData(exp5.subset)$median.reads == 0,]
```

```{r}
exp5.subset <- exp5.subset[rowData(exp5.subset)$median.reads > 0,] # 2 probes filtered for exp 5
```

# Normalization

```{r}
exp5.subset <- calcNormCounts(exp5.subset)
```

## Copy Number Calculation

Rename cluster labels

```{r}
colData(exp5.subset)$cluster <- fct_recode(colData(exp5.subset)$cluster, HCEC = "1", RPE1 = "2")
colData(exp5.subset)$cluster <- fct_drop(colData(exp5.subset)$cluster)
fct_count(colData(exp5.subset)$cluster)
```

Following probes have control cell median equal to 0. Filter out prior to proceeding.

```{r}
rpe1.bcs <- colData(exp5.subset) %>% as_tibble() %>% filter(cluster == "RPE1") %>% pull(cell.barcode)
names(which(apply(assay(exp5.subset[,rpe1.bcs], "normcounts"), 1, median) == 0))

amps.to.exlcude <- c("CO610_AMP155", "CO610_AMP263")
exp5.subset <- exp5.subset[!rownames(exp5.subset) %in% amps.to.exlcude,]
```

Generate control.copy.number lookup table for RPE1 cells, set 10q to ploidy = 3

```{r}
control.copy.number <- generateControlCopyNumberTemplate(exp5.subset, sample.feature.label = "RPE1", copy.number = 2)
control.copy.number["chr10q", "copy.number"] <- 3
exp5.subset <- calcCopyNumber(exp5.subset, control.copy.number = control.copy.number, sample.feature = "cluster")
```

```{r}
cn.tidy <- getTidyData(exp5.subset, assay = "copyNumber")
```

```{r}
cn.tidy %>%
  group_by(feature.id, cluster) %>%
  summarize(median.ploidy = median(copyNumber), .groups = "drop") %>%
  ggplot(aes(x = feature.id, y = median.ploidy)) +
  geom_point(size = 1) +
  facet_wrap(cluster ~ ., ncol = 1) +
  theme(axis.text.x = element_blank()) +
    ylim(0,8)
```

## Copy Number Smoothing

```{r}
exp5.subset <- calcSmoothCopyNumber(exp5.subset)
rpe1.bc <- colData(exp5.subset) %>% as_tibble() %>% filter(cluster == "RPE1") %>% pull(cell.barcode)
exp5.subset <- calcGMMCopyNumber(exp5.subset, cell.barcodes = rpe1.bc, control.copy.number = control.copy.number, model.components = 1:4)
```

```{r}
smoothed.cn.tidy.chr <- getTidyData(exp5.subset, alt.exp = "smoothedCopyNumberByChr", assay = "smoothedCopyNumber")
```

```{r}
ggplot(smoothed.cn.tidy.chr, aes(x = feature.id, y = smoothedCopyNumber)) + 
    facet_wrap(cluster ~., ncol = 1) +
    ggdist::stat_halfeye(point_size = 2, alpha = 0.8)
```

```{r}
smoothed.cn.tidy.arm <- getTidyData(exp5.subset, alt.exp = "smoothedCopyNumberByArm", assay = "smoothedCopyNumber")
```

```{r}
ggplot(smoothed.cn.tidy.arm, aes(x = feature.id, y = smoothedCopyNumber)) +
  facet_wrap(cluster ~ ., ncol = 1) +
  ggdist::stat_halfeye(point_size = 2, alpha = 0.8) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  scale_y_continuous(breaks = c(0, 2, 4, 6, 8, 10))
```

# Heatmaps

```{r}
assayHeatmap(exp5.subset, assay = "copyNumber", split.col.by = "chr", split.row.by = "cluster", color.preset = "copy.number.denoise", name = "CN")
```

```{r}
assayHeatmap(exp5.subset, alt.exp = "smoothedCopyNumberByChr", assay = "smoothedCopyNumber", split.row.by = "cluster", color.preset = "copy.number.denoise", name = "smCN", column_names_rot = 0, column_names_centered = T)
```

```{r}
assayHeatmap(exp5.subset, alt.exp = "smoothedCopyNumberByArm", assay = "smoothedCopyNumber", split.row.by = "cluster", color.preset = "copy.number.denoise", name = "smCN")
```

```{r}
assayHeatmap(exp5.subset, alt.exp = "smoothedCopyNumberByChr", assay = "gmmCopyNumber", split.row.by = "cluster", color.preset = "copy.number.denoise", name = "gmmCN", column_names_rot = 0, column_names_centered = T)
```

```{r}
assayHeatmap(exp5.subset, alt.exp = "smoothedCopyNumberByArm", assay = "gmmCopyNumber", split.row.by = "cluster", color.preset = "copy.number.denoise", name = "gmmCN")
```

# Save Objects

```{r}
saveRDS(exp5, file = here("datasets", "exp05.full.tapestriexperiment.RDS"))
saveRDS(exp5.subset, file = here("datasets", "exp05.tapestriexperiment.RDS"))
```

# session_info

```{r}
sessioninfo::session_info()
```
