---
title: "Figure L - Exp10 Sensitivity Metrics"
author: "Joey Mays"
date: '2024-04-30'
output:
  pdf_document:
    toc: yes
  html_notebook:
    code_folding: show
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
    theme: flatly
    df_print: paged
  html_document:
    code_folding: show
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
    theme: flatly
    df_print: paged
editor_options:
  chunk_output_type: inline
  markdown:
    wrap: 72
---

# Setup

```{r message=FALSE}
library(karyotapR)
library(ggplot2)
library(here)
library(cowplot)
library(ggrastr)
library(tidyverse)
library(ComplexHeatmap)
```

```{r}
set.seed(20240430)
```

```{r}
here::i_am("analysis-notebooks/Figure-L_exp10_sensitivity.Rmd")
```

```{r}
source(here("scripts", "gmm-calc-TPR.R"))
```

```{r}
dir.create(here("figures", "figure-l"), showWarnings = F)
fig.scalar = 1.375

theme_bw_custom <- theme_bw()
theme_bw_custom$axis.text <- element_text(size = 7)
theme_bw_custom$axis.title <- element_text(size = 8)
theme_bw_custom$legend.text <- element_text(size = 7)
theme_bw_custom$legend.title <- element_text(size = 8)
theme_bw_custom$strip.text <- element_text(size = 7)

theme_classic_custom <- theme_classic()
theme_classic_custom$axis.text <- element_text(size = 7)
theme_classic_custom$axis.title <- element_text(size = 8)
theme_classic_custom$legend.text <- element_text(size = 7)
theme_classic_custom$legend.title <- element_text(size = 8)
theme_classic_custom$strip.text <- element_text(size = 7)
```

# Load Data

```{r}
exp10 <- readRDS(file = here("datasets", "exp10.tapestriexperiment.RDS"))
exp10.full <- readRDS(file = here("datasets", "exp10.full.tapestriexperiment.RDS"))
```

Move RPE1 to first position in factor for plotting order.

```{r}
colData(exp10)$cluster <- fct_relevel(colData(exp10)$cluster , "RPE1")
```

# Allele Freq UMAP

```{r}
scatter.raw <- reducedDim(altExp(exp10.full, "alleleFrequency"), "UMAP")
scatter.raw$cluster <- colData(exp10.full)$cluster
```

```{r}
scatter.raw$cluster <- fct_relevel(scatter.raw$cluster, "RPE1") #Move RPE1 to first position in factor for plotting order.
```

```{r}
exp10.umap <- 
    ggplot(data = scatter.raw, aes(x = umap.1, y = umap.2, color = cluster)) +
    geom_point(size = 0.3, alpha = 0.7) +
    theme_bw_custom + 
    theme(legend.position = "top", axis.text = element_blank(), axis.ticks = element_blank(), 
          panel.grid = element_blank(), legend.box.margin = margin(0,0,0,0, "pt"), 
          plot.margin = unit(c(0,5.5,5.5,5.5), "pt"), legend.margin = margin(0,0,0,0,"pt")) +
    labs(title = "", x = "umap 1", y = "umap 2", fill = "", color = "") +
    scale_color_manual(values = c("#785EF0", "#648FFF", "#DC267F", "gray"))
exp10.umap
```

```{r}
saveRDS(exp10.umap, here("figures", "figure-l", "exp10.af.umap.RDS"))
```

```{r}
#Rasterize points so Illustrator doesn't complain
rasterize(exp10.umap, layers = "point", dpi = 300)
```

```{r}
pdf(file = here("figures", "figure-l", "exp10.af.umap.pdf"), 
    width = 3.2, height = 2.0)
#Rasterize points so Illustrator doesn't complain
#rasterize(exp10.umap, layers = "Point", dpi = 1200)
exp10.umap
dev.off()
```

# Y Chromosome Counts

```{r}
y.data <- getTidyData(exp10, alt.exp = "chrYCounts")

ychr.plot <- 
    ggplot(y.data, aes(x=feature.id, y = counts, color = cluster)) +
    geom_point(position = position_jitter(width = 0.3), size = 0.5) +
    facet_wrap(cluster ~ .) +
    theme_bw_custom + 
    theme(strip.background = element_blank(), axis.text.x = element_text(hjust = 1, angle = 45), panel.grid.minor = element_blank(), legend.position = "none") +
    labs(x = "Chr Y Probe ID", y = "Read Counts") +
    scale_y_continuous(expand = c(0,0,0,0)) +
    scale_color_manual(values = c("#785EF0", "#648FFF", "#DC107F"))
ychr.plot
```

```{r}
saveRDS(ychr.plot, here("figures", "figure-l", "chrY.probes.RDS"))
```

# Smoothed CN Heatmap

## Whole Chromosome

```{r}
row.annotation.data <- getTidyData(exp10, assay = "copyNumber") %>%
      dplyr::select("cell.barcode", "cluster") %>%
      dplyr::distinct() %>%
      dplyr::pull(cluster) %>%
      tibble::enframe(name = NULL, value = "cluster") %>%
      as.data.frame()

color.vector <- c("#785EF0", "#648FFF", "#DC267F")
names(color.vector) <- c("RPE1", "LoVo", "LS513")
color.list <- list(color.vector)
names(color.list)[1] <- "cluster"
row.annotation <- ComplexHeatmap::rowAnnotation(
  df = row.annotation.data, col = color.list, border = TRUE, na_col = "white",
  annotation_name_side = "top", annotation_name_gp = grid::gpar(fontsize = 0), show_legend = FALSE
)
```

```{r}
bottom.anno.scale <- c("1" = "#2c7bb6", "2" = "#ffffff", "3" = "#fdae61", "4" = "#d7191c")

bottom.annotation <- ComplexHeatmap::columnAnnotation(cell1 = c(1,2,3,4, rep(2,19)),
                                                      col = list(cell1 = bottom.anno.scale),  annotation_name_side = "left", 
                                                      annotation_name_gp = grid::gpar(fontsize = 0), border = TRUE)
```

```{r}
exp10.wc.smoothed.heatmap <- assayHeatmap(exp10, alt.exp = "smoothedCopyNumberByChr", assay = "smoothedCopyNumber", split.col.by = "feature.id", split.row.by = "cluster", annotate.row.by = "cluster", column_title_rot = 0, color.custom = circlize::colorRamp2(c(1, 1.5, 2, 2.5, 3, 4), c("#2c7bb6", "#ffffff", "#ffffff", "#ffffff", "#fdae61", "#d7191c")), left_annotation = row.annotation, heatmap_legend_param = list(title_position = "lefttop-rot", border = "black"), name = "Copy Number")
exp10.wc.smoothed.heatmap
```

```{r}
pdf(file = here("figures", "figure-l", "exp10.wc.smoothed.heatmap.pdf"), 
    width = 7.90, height = 3.75)
draw(exp10.wc.smoothed.heatmap, use_raster = T, raster_quality = 7)
dev.off()
```

### GMM Heatmap

```{r}
gmm.anno.scale <- c("1" = "#2c7bb6", "2" = "#ffffff", "3" = "#fdae61", "4" = "#d7191c", "5" = "#d7191c", "6" = "#d7191c")
```

```{r}
exp10.wc.gmm.heatmap <- 
    assayHeatmap(exp10, alt.exp = "smoothedCopyNumberByChr", assay = "gmmCopyNumber", split.col.by = "feature.id", split.row.by = "cluster", column_title_rot = 0, color.custom = gmm.anno.scale, heatmap_legend_param = list(title_position = "topcenter", border = "black", labels = c("4+", "3", "2", "1"), at = 4:1), name = "Single Cell\nCopy Number\nCall")
exp10.wc.gmm.heatmap
```

```{r}
saveRDS(exp10.wc.gmm.heatmap, file = here("figures", "figure-l", "exp10.wc.gmm.heatmap.RDS"))
```

```{r}
pdf(file = here("figures", "figure-l", "exp10.wc.gmm.heatmap.pdf"), 
    width = 7.90, height = 3.75)
ht_opt("HEATMAP_LEGEND_PADDING" = unit(10, "mm"))
draw(exp10.wc.gmm.heatmap, use_raster = T, raster_quality = 7, padding = unit(c(2,2,2,8), "mm"))
ht_opt("HEATMAP_LEGEND_PADDING" = unit(10, "mm"))
dev.off()
```

### Raw Matrix

```{r}
raw.ploidy.wc <- round(t(assay(altExp(exp10, "smoothedCopyNumberByChr"), "smoothedCopyNumber")), 3)
colnames(raw.ploidy.wc) <- paste0("chr", colnames(raw.ploidy.wc))
raw.ploidy.wc <- cbind(colData(exp10)[,"cluster", drop = F], raw.ploidy.wc) 
raw.ploidy.wc <- raw.ploidy.wc %>% as.data.frame() %>% rownames_to_column("cell.barcode")
```

```{r, eval = F}
write.csv(x = raw.ploidy.wc, file = here("figures", "figure-l", "exp10.wc.smooth.heatmap.raw.csv"), 
          quote = F, row.names = F)
```

## Chromosome Arms

```{r}
bottom.anno.scale <- c("1" = "#2c7bb6", "2" = "#ffffff", "3" = "#fdae61", "4" = "#d7191c")

bottom.annotation.arm <- ComplexHeatmap::columnAnnotation(cell1 = c(1,2,3,4, rep(2,37)),
                                                      col = list(cell1 = bottom.anno.scale),  
                                                      annotation_name_side = "left", annotation_name_gp = grid::gpar(fontsize = 0), 
                                                      border = TRUE)
```

```{r}
exp10.arm.smoothed.heatmap <- assayHeatmap(exp10, alt.exp = "smoothedCopyNumberByArm", assay = "smoothedCopyNumber", split.col.by = "feature.id", split.row.by = "cluster", annotate.row.by = "cluster", column_title_rot = 90, color.custom = circlize::colorRamp2(c(1, 1.5, 2, 2.5, 3, 4), c("#2c7bb6", "#ffffff", "#ffffff", "#ffffff", "#fdae61", "#d7191c")), left_annotation = row.annotation, heatmap_legend_param = list(title_position = "lefttop-rot", border = "black"), name = "Copy Number")
exp10.arm.smoothed.heatmap
```

```{r}
pdf(file = here("figures", "figure-l", "exp10.arm.smoothed.heatmap.pdf"), 
    width = 7.90, height = 3.75)
draw(exp10.arm.smoothed.heatmap, use_raster = T, raster_quality = 7)
dev.off()
```

### GMM Heatmap

```{r}
exp10.hm <- exp10
#relabel arms for heatmap
rownames(altExp(exp10.hm, "smoothedCopyNumberByArm")) <- str_split_i(rownames(altExp(exp10.hm, "smoothedCopyNumberByArm")), "chr", 2)
```

```{r}
gmm.anno.scale <- c("1" = "#2c7bb6", "2" = "#ffffff", "3" = "#fdae61", "4" = "#d7191c", "5" = "#d7191c", "6" = "#d7191c")
```

```{r}
exp10.arm.gmm.heatmap <- 
    assayHeatmap(exp10.hm, alt.exp = "smoothedCopyNumberByArm", assay = "gmmCopyNumber", split.col.by = "feature.id", split.row.by = "cluster", column_title_rot = 45, color.custom = gmm.anno.scale, heatmap_legend_param = list(title_position = "topcenter", border = "black", labels = c("4+", "3", "2", "1"), at = 4:1), name = "Single Cell\nCopy Number\nCall")
exp10.arm.gmm.heatmap
```

```{r}
saveRDS(exp10.arm.gmm.heatmap, file = here("figures", "figure-l", "exp10.arm.gmm.heatmap.RDS"))
```


```{r}
pdf(file = here("figures", "figure-l", "exp10.arm.gmm.heatmap.pdf"), 
    width = 7.90, height = 3.75)
draw(exp10.arm.gmm.heatmap, use_raster = T, raster_quality = 7)
dev.off()
```

### Raw Matrix

```{r}
raw.ploidy.arm <- round(t(assay(altExp(exp10, "smoothedCopyNumberByArm"), "smoothedCopyNumber")), 3)
raw.ploidy.arm <- cbind(colData(exp10)[,"cluster", drop = F], raw.ploidy.arm) 
raw.ploidy.arm <- raw.ploidy.arm %>% as.data.frame() %>% rownames_to_column("cell.barcode")
```

```{r, eval = F}
write.csv(x = raw.ploidy.arm, file = here("figures", "figure-l", "exp10.arm.smooth.heatmap.raw.csv"), 
          quote = F, row.names = F)
```

# Sensitivity

## Whole Chromosome

### Bar Chart

```{r}
rpe1.data <- getTidyData(exp10, "smoothedCopyNumberByChr", "gmmCopyNumber") %>% filter(cluster == "RPE1", feature.id != 10) %>% select(!c("total.reads", "cluster"))
rpe1.data$known.ploidy <- 2
rpe1.data.results <- rpe1.data %>% group_by(cell.barcode, feature.id) %>% summarise(ploidy.concordance = gmmCopyNumber == known.ploidy, .groups = "drop_last") %>% group_by(feature.id) %>% dplyr::count(ploidy.concordance) %>% mutate(pct.concordance = round(n*100/sum(n), 1)) %>% filter(ploidy.concordance == TRUE) %>% print()
```

```{r}
sum(rpe1.data.results$pct.concordance >=70.0)
sum(rpe1.data.results$pct.concordance >=75.0)
nrow(rpe1.data.results)
```

```{r}
rpe1.wc.sensitivity <- 
rpe1.data.results %>% 
    ggplot(aes(x=feature.id, y=pct.concordance)) +
    geom_col(color = "black", fill = "#C6DEA6", width = .75) +
    coord_cartesian(ylim = c(0,100)) +
    scale_y_continuous(expand = c(0,0,0,5), breaks = seq(0,100,20)) +
    labs(y = "Accuracy (%)", x = "Chromosome") +
    geom_text(aes(label=round(pct.concordance,0)), vjust = -1, size = 7/.pt) +
    theme_bw_custom +
    theme(panel.grid.major.x = element_blank(), panel.grid.minor = element_blank(), panel.border = element_blank(), axis.line = element_line()) +
    geom_hline(yintercept = mean(rpe1.data.results$pct.concordance), linetype = "dashed", alpha = 0.7) +
    annotate("text", x = 21, y = 102, size = 7/.pt, label = paste0("avg: ", round(mean(rpe1.data.results$pct.concordance), 0), "%"))
rpe1.wc.sensitivity
```

```{r}
saveRDS(rpe1.wc.sensitivity, here("figures", "figure-l", "exp10.rpe1.sensitivity.wc.RDS"))
```

### Panel v2 Comparison

```{r}
n.probes <- fct_count(rowData(exp10)$chr)
colnames(n.probes) <- c("feature.id", "n.probes")
rpe1.data.results <- rpe1.data.results %>% left_join(n.probes, by = "feature.id")
```

```{r}
exp4.wc.results <- readRDS(here("datasets", "exp04.rpe1.accuracy.rates.RDS"))
```

```{r}
combined.results <- rpe1.data.results %>% inner_join(exp4.wc.results %>% select(feature.id, pct.concordance, n.probes), by = "feature.id", suffix = c(".v3", ".v2")) %>% select(!c("ploidy.concordance", "n"))
```

```{r}
combined.results <- combined.results %>% mutate(delta.acc = pct.concordance.v3 - pct.concordance.v2, delta.probes = n.probes.v3 - n.probes.v2)
```

```{r}
rpe1.wc.sensitivity.change <- 
ggplot(combined.results, aes(x = feature.id, y = delta.acc)) +
    geom_col(fill = "#C6DEA6", color = "black", width = .75) +
    theme_bw_custom +
    theme(panel.border = element_blank(), axis.line = element_line(), panel.grid.minor.y = element_blank()) +
    labs(x = "Chromosome", y = "\u0394 Accuracy (pp)") +
    scale_y_continuous(limits = c(-10, 10), expand = c(0,0,0,0)) +
    geom_text(aes(y = 8.5, label = sprintf("%+2d", delta.probes)), size = 7/.pt) +
    geom_hline(yintercept = 0)
rpe1.wc.sensitivity.change
```

```{r}
saveRDS(rpe1.wc.sensitivity.change, here("figures", "figure-l", "exp10.rpe1.sensitivity.wc.change.RDS"))
```

```{r}
change.rsq <- summary(lm(data = combined.results, delta.acc ~ delta.probes)) %>% print() %>% .$r.squared
change.cor <- cor(combined.results$delta.probes, combined.results$delta.acc) %>% print()
```

```{r}
rpe1.wc.sensitivity.change.lm <- 
ggplot(combined.results, aes(x = delta.probes, y = delta.acc)) +
    geom_smooth(method = "lm", se = F, color = "#C6DEA6") + 
    geom_point(alpha = 0.5, size = 2) + 
    theme_bw_custom + 
    theme(panel.border = element_blank(), axis.line = element_line(), panel.grid.minor.y = element_blank()) +
    labs(x = "\u0394 Probe Number", y = "\u0394 Accuracy (pp)") +
    scale_x_continuous(breaks = seq(-6, 8, 2)) +
    scale_y_continuous(limits = c(-10, 30), expand = c(0,0,0,0)) +
    annotate("text", x = -5, y = 28, size = 7/.pt, hjust = 0, label = paste0("Pearson r = ", round(change.cor, 2))) +
    annotate("text", x = -5, y = 25, size = 7/.pt, hjust = 0, label = paste0("R-sq = ", round(change.rsq, 2)))
rpe1.wc.sensitivity.change.lm
```

```{r}
#saveRDS(rpe1.wc.sensitivity.change.lm, here("figures", "figure-l", "exp10.rpe1.sensitivity.wc.change.regression.RDS"))
```

### Theoretical Sensitivity

```{r}
exp4.gmm.metrics <- readRDS(here("datasets", "exp4.gmm.metrics.RDS"))
```

```{r}
exp10.gmm.metrics <- calcTruePosRate(exp10, "chr") %>% filter(cn.sim.class != "sim_cn6")
exp10.gmm.metrics <- exp10.gmm.metrics %>% left_join(n.probes, by = "feature.id")
exp10.gmm.metrics <- exp10.gmm.metrics %>% mutate(false.neg.p = false.neg.p * 100, true.pos.p = true.pos.p * 100)
```

```{r}
tpr.wc.bars <- 
ggplot(exp10.gmm.metrics, aes(x = feature.id, y = true.pos.p)) +
    geom_col(aes(fill = cn.sim.class), color = "black", width = 0.85) +
    facet_wrap(cn.sim.class ~., ncol = 1) +
    theme_classic_custom +
    theme(legend.position = "none", panel.grid.major = element_line(), 
          strip.background = element_blank()) +
    scale_fill_manual(values = c("#785EF0", "#648FFF", "#DC267F", "#FE6100", "#FFB000")) +
    labs(x = "Chromosome", y = "Theoretical Sensitivity (%)") +
    scale_y_continuous(breaks = c(0, 50, 100), expand = c(0,0,0,10)) +
    geom_hline(yintercept = 0) +
    geom_text(aes(label=round(true.pos.p,0)), vjust = -1, size = 6/.pt, angle = 45, hjust = 0)
tpr.wc.bars
```

```{r}
#saveRDS(tpr.wc.bars, here("figures", "figure-l", "exp10.rpe1.tpr.wc.RDS"))
```

```{r}
tpr.wc.nprobes <- 
ggplot(exp10.gmm.metrics, aes(x=n.probes, y=true.pos.p, color = cn.sim.class)) +
    geom_point(alpha = 0.5, position = position_jitter(width = 0.05)) + 
    #geom_point(data = rpe1.data.results, aes(x=n.probes, y=pct.concordance), alpha = 0.5, inherit.aes = F) + 
    theme_classic_custom +
    theme(legend.position = "top", panel.grid.major = element_line()) +
    theme(legend.box.margin = margin(0, 10, 0, 10), legend.margin = margin(0), plot.margin = unit(c(0, 5.5, 5.5, 5.5), "pt")) +
    geom_smooth(se = F, span = 1) +
    geom_smooth(data = rpe1.data.results, aes(x=n.probes, y=pct.concordance), 
                color = "black", inherit.aes = F, se = F, span = 1, linetype = "dashed") +
    scale_color_manual(values = viridisLite::viridis(n=5), labels = 1:5) +
    labs(x = "Number of Probes", y = "Theoretical Sensitivity (%)", color = "Copy Number") +
    scale_x_continuous(limits = c(10,26), breaks = seq(10,26,2))
tpr.wc.nprobes
```

Correlation between theoretical true positive rate and actual true positive rate is high.

```{r}
combined.data <- inner_join(exp4.gmm.metrics, rpe1.data.results, by = "feature.id") %>% filter(cn.sim.class == "sim_cn2")
cor(combined.data$pct.concordance, combined.data$true.pos.p)
```

```{r}
sensitivity.compare <- inner_join(exp4.gmm.metrics %>% select(feature.id, cn.sim.class, true.pos.p), 
           exp10.gmm.metrics %>% select(feature.id, cn.sim.class, true.pos.p),
           by = c("feature.id", "cn.sim.class"), suffix = c(".exp4", ".exp10")) %>% 
    mutate(delta.tpr = true.pos.p.exp10 - true.pos.p.exp4) %>% 
    inner_join(combined.results %>% select(feature.id, delta.probes), by = "feature.id")
```

```{r}
tpr.wc.bars.compare <- 
    ggplot(sensitivity.compare, aes(x = feature.id, y = delta.tpr)) +
    geom_col(aes(fill = cn.sim.class), color = "black", width = 0.85) +
    facet_wrap(cn.sim.class ~., ncol = 1) +
    theme_bw_custom +
    theme(legend.position = "none", panel.grid.major = element_line(), 
          strip.background = element_blank()) +
    scale_fill_viridis_d() +
    labs(x = "Chromosome", y = "Theoretical Sensitivity (%)") +
    scale_y_continuous(limits = c(-10, 30), breaks = c(-10, 0, 10, 20, 30), expand = c(0,0,0,10)) +
    geom_text(aes(y = 30, label = sprintf("%+2d",delta.probes)), size = 3)
tpr.wc.bars.compare
```

```{r}
sensitivity.compare.avg <- sensitivity.compare %>% group_by(cn.sim.class) %>% summarize(mean.tpr4 = mean(true.pos.p.exp4), mean.tpr10 = mean(true.pos.p.exp10))
sensitivity.compare.avg$cn.sim.class.label <- paste0("Copy Number = ", 1:5)
```

```{r}
sensitivity.compare.plot <- 
ggplot(sensitivity.compare.avg, aes(x = mean.tpr4, y = mean.tpr10, color = cn.sim.class, label = cn.sim.class.label)) +
    geom_point() +
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "red") + 
    theme_classic_custom +
    theme(panel.grid.major = element_line(), panel.grid.minor = element_line()) + 
    scale_x_continuous(limits = c(35, 100), expand = c(0,0,0,0)) + 
    scale_y_continuous(limits = c(35, 100), expand = c(0,0,0,0)) +
    geom_text(nudge_x = -4, nudge_y = 1, size = 6/.pt, color = "black") +
    guides(color = "none") + 
    scale_color_manual(values = c("#785EF0", "#648FFF", "#DC267F", "#FE6100", "#FFB000")) +
    labs(x = "Panel Version 3\nMean Theoretical Sensitivity (%)", y = "Panel Version 2\nMean Theoretical Sensitivity (%)")
sensitivity.compare.plot
``` 

```{r}
saveRDS(sensitivity.compare.plot, here("figures", "figure-l", "exp10.rpe1.sensitivity.compare.RDS"))
```

### Cross Validation

```{r}
rpe1.bcs <- colData(exp10) %>% as_tibble %>% filter(cluster == "RPE1") %>% pull(cell.barcode)
rpe1.bcs.scrambled <- sample(rpe1.bcs, size = length(rpe1.bcs), replace = F)
rpe1.index <- cut_number(seq_along(rpe1.bcs.scrambled), n = 5)
rpe1.index <- rpe1.index %>% fct_recode("fold1" = "[1,182]", "fold2" = "(182,364]", "fold3" = "(364,545]", "fold4" = "(545,727]", "fold5" = "(727,908]")
cv.folds <- data.frame(cell.barcode = rpe1.bcs.scrambled, cv.fold = rpe1.index)
```

```{r}
exp10.cv <- exp10
control.copy.number <- generateControlCopyNumberTemplate(exp10, copy.number = 2, "RPE1")
control.copy.number["chr10q", "copy.number"] <- 3
```

```{r}
cv.results <- map(levels(cv.folds$cv.fold), function(x){
    
    training.bcs <- cv.folds %>% filter(cv.fold != x) %>% pull(cell.barcode)
    exp10.cv <- calcGMMCopyNumber(exp10.cv, cell.barcodes = training.bcs, control.copy.number = control.copy.number)
    
    rpe1.cv.data <- getTidyData(exp10.cv, "smoothedCopyNumberByChr", "gmmCopyNumber") %>% filter(cluster == "RPE1", feature.id != 10, !cell.barcode %in% training.bcs) %>% select(!c("total.reads", "cluster"))
    rpe1.cv.data$known.ploidy <- 2
    rpe1.cv.data.results <- rpe1.cv.data %>% group_by(cell.barcode, feature.id) %>% summarise(cn.concordance = gmmCopyNumber == known.ploidy, .groups = "drop") %>% group_by(feature.id) %>% dplyr::count(cn.concordance) %>% mutate(pct.concordance = round(n*100/sum(n), 1)) %>% filter(cn.concordance == TRUE) %>% add_column(test.fold = x)
    
}) %>% list_rbind() %>% left_join(n.probes, by = "feature.id")
```

```{r}
cv.results %>% 
    ggplot(aes(x = feature.id, y = pct.concordance)) +
    ggdist::stat_pointinterval(.width = c(1), linewidth = 2, point_size = 1, point_interval = "mean_qi") + 
    geom_point(position = position_nudge(x=-0.3), alpha = 0.5, size = 1) +
    geom_point(data = rpe1.data.results, aes(x=feature.id, y=pct.concordance), inherit.aes = F, pch = 4, color = "red", position = position_nudge(x=0.3)) +
    theme_bw_custom +
    scale_y_continuous(limits = c(34, 100)) +
    labs(y = "Accuracy (%)", x = "Chromosome")
```

```{r}
wc.sensitivity.mad <- 
cv.results %>% group_by(feature.id) %>% summarize(mean.pct = mean(pct.concordance), mad = mad(pct.concordance), mean.ad = DescTools::MeanAD(pct.concordance)) %>% print() %>% 
    ggplot(aes(x = feature.id, y = mean.pct)) + 
    ggdist::geom_pointinterval(aes(ymin = mean.pct - mean.ad, ymax = mean.pct + mean.ad), shape = 16) +
    theme_classic_custom + 
    theme(panel.grid.major = element_line()) +
    scale_y_continuous(limits = c(60, 100), expand = c(0,0,0,0)) +
    labs(y = "Accuracy (%)", x = "Chromosome") + 
    geom_hline(yintercept = mean(cv.results$pct.concordance), linetype = "dashed", alpha = 0.7) +
    annotate("text", x = 21, y = 95, size = 7/.pt, label = paste0("avg: ", round(mean(cv.results$pct.concordance), 0), "%"))
wc.sensitivity.mad
```
```{r}
saveRDS(wc.sensitivity.mad, here("figures", "figure-l", "exp10.rpe1.sensitivity.wc.mad.RDS"))
```

### 3 Component Model

```{r}
control.copy.number <- generateControlCopyNumberTemplate(exp10, copy.number = 2, "RPE1")
control.copy.number["chr10q", "copy.number"] <- 3
exp10.3comp <- calcGMMCopyNumber(exp10, cell.barcodes = rpe1.bcs, control.copy.number = control.copy.number, model.components = 1:3)
exp10.gmm.metrics.3 <- calcTruePosRate(exp10.3comp, "chr") %>% filter(cn.sim.class != "sim_cn6")
exp10.gmm.metrics.3 <- exp10.gmm.metrics.3 %>% left_join(n.probes, by = "feature.id")
exp10.gmm.metrics.3 <- exp10.gmm.metrics.3 %>% mutate(false.neg.p = false.neg.p * 100, true.pos.p = true.pos.p * 100)
```

```{r}
exp10.gmm.metrics.3.avg <- exp10.gmm.metrics.3 %>% group_by(cn.sim.class) %>% summarize(mean.tpr = mean(true.pos.p), .groups = "drop") %>% print()
```

```{r}
exp10.3comp.wc.sens <- 
ggplot(exp10.gmm.metrics.3, aes(x = feature.id, y = true.pos.p)) +
    geom_col(aes(fill = cn.sim.class), color = "black", width = 0.85) +
    facet_wrap(cn.sim.class ~., ncol = 1) +
    theme_classic_custom +
    theme(legend.position = "none", panel.grid.major.y = element_line(), 
          strip.background = element_blank()) +
    scale_fill_manual(values = c("#785EF0", "#648FFF", "#DC267F")) +
    labs(x = "Chromosome", y = "Theoretical Sensitivity (%)") +
    scale_y_continuous(breaks = c(0, 50, 100), expand = c(0,0,0,0)) +
    geom_text(aes(label=round(true.pos.p,0)), vjust = -1, size = 6/.pt) +
    geom_text(data = exp10.gmm.metrics.3.avg, aes(label = glue::glue("avg: {round(mean.tpr, 0)}%"), x = 22, y = 140), size = 6/.pt, inherit.aes = F)
exp10.3comp.wc.sens
```

```{r}
#saveRDS(exp10.3comp.wc.sens, here("figures", "figure-l", "exp10.rpe1.tpr.wc.3comp.RDS"))
```

```{r}
ggplot(exp10.gmm.metrics.3, aes(x=n.probes, y=true.pos.p, color = cn.sim.class)) +
    geom_point(alpha = 0.5, position = position_jitter(width = 0.25)) + 
    theme_classic_custom +
    theme(legend.position = "top", panel.grid.major = element_line()) +
    theme(legend.box.margin = margin(0, 10, 0, 10), legend.margin = margin(0), plot.margin = unit(c(0, 5.5, 5.5, 5.5), "pt")) +
    geom_smooth(se = F, span = 1, linewidth = 0.75) +
    geom_smooth(data = rpe1.data.results, aes(x=n.probes, y=pct.concordance), linewidth = 0.75,
                color = "black", inherit.aes = F, se = F, span = 1, linetype = "dashed", alpha = 0.5) +
    scale_color_manual(values = c("#785EF0", "#648FFF", "#DC267F"), labels = 1:3) +
    labs(x = "Number of Probes", y = "Theoretical Sensitivity (%)", color = "Copy Number")
```

```{r}
exp10.gmm.metrics.3.spec <- calcTrueNegRate(exp10.3comp) %>% mutate(tnr = tnr * 100)
exp10.gmm.metrics.3.spec <- exp10.gmm.metrics.3.spec %>% left_join(n.probes, by = "feature.id")

ggplot(exp10.gmm.metrics.3.spec, aes(x = feature.id, y = tnr)) +
    geom_col(aes(fill = cn.sim.class), color = "black", width = 0.85) +
    facet_wrap(cn.sim.class ~., ncol = 1) +
    theme_classic_custom +
    theme(legend.position = "none", panel.grid.major.y = element_line(), 
          strip.background = element_blank()) +
    scale_fill_manual(values = c("#785EF0", "#648FFF", "#DC267F")) +
    labs(x = "Chromosome", y = "Theoretical Specificity (%)") +
    scale_y_continuous(breaks = c(0, 50, 100), expand = c(0,0,0,25)) +
    geom_text(aes(label=round(tnr,0)), vjust = -1, size = 6/.pt)
```

```{r}
ggplot(exp10.gmm.metrics.3.spec, aes(x=n.probes, y=tnr, color = cn.sim.class)) +
    geom_point(alpha = 0.5, position = position_jitter(width = 0.25)) + 
    theme_classic_custom +
    theme(legend.position = "top", panel.grid.major = element_line()) +
    theme(legend.box.margin = margin(0, 10, 0, 10), legend.margin = margin(0), plot.margin = unit(c(0, 5.5, 5.5, 5.5), "pt")) +
    geom_smooth(se = F, span = 1, linewidth = 0.75) +
    scale_color_manual(values = c("#785EF0", "#648FFF", "#DC267F"), labels = 1:5) +
    labs(x = "Number of Probes", y = "Theoretical Specificity (%)", color = "Copy Number")
```

## Chr Arms

### Bar Chart

```{r}
rpe1.data.arm <- getTidyData(exp10, "smoothedCopyNumberByArm", "gmmCopyNumber") %>% filter(cluster == "RPE1") %>% select(!c("total.reads", "cluster"))
rpe1.data.arm$known.ploidy <- 2
rpe1.data.arm[which(rpe1.data.arm$feature.id == "chr10q"), "known.ploidy"] <- 3

rpe1.data.arm.results <- rpe1.data.arm %>% group_by(cell.barcode, feature.id) %>% summarise(ploidy.concordance = gmmCopyNumber == known.ploidy) %>% group_by(feature.id) %>% dplyr::count(ploidy.concordance) %>% mutate(pct.concordance = round(n*100/sum(n), 1)) %>% filter(ploidy.concordance == TRUE) %>% print()
```

```{r}
sum(rpe1.data.arm.results$pct.concordance >=75.0)
nrow(rpe1.data.arm.results)
```

```{r}
rpe1.data.arm.results$feature.label <- rpe1.data.arm.results$feature.id %>% str_sub(4,-1)
```

```{r}
rpe1.arm.sensitivity <-
rpe1.data.arm.results %>% 
    ggplot(aes(x=feature.id, y=pct.concordance)) +
    geom_col(color = "black", fill = "#7EBDC3", width = 0.75) +
    coord_cartesian(ylim = c(0,100)) +
    scale_y_continuous(expand = c(0,0), breaks = seq(0,100,20)) +
    labs(y = "Accuracy (%)", x = "Chromosome Arm") +
    geom_text(aes(label=round(pct.concordance,0)), vjust = -1, size = 6/.pt) +
    theme_bw_custom +
    theme(panel.grid.major.x = element_blank(), panel.grid.minor = element_blank(), axis.text.x = element_text(hjust = 1, angle = 45), axis.line = element_line(), panel.border = element_blank()) +
    geom_hline(yintercept = mean(rpe1.data.arm.results$pct.concordance), linetype = "dashed", alpha = 0.7) +
    annotate("text", x = 39, y = 95, size = 7/.pt, label = paste0("avg: ", round(mean(rpe1.data.arm.results$pct.concordance), 0), "%")) +
    scale_x_discrete(labels = rpe1.data.arm.results$feature.label)
rpe1.arm.sensitivity
```

```{r}
saveRDS(rpe1.arm.sensitivity, here("figures", "figure-l", "exp10.rpe1.accuracy.arm.RDS"))
```

### Panel v2 Comparison

```{r}
n.probes.arm <- fct_count(rowData(exp10)$arm)
colnames(n.probes.arm) <- c("feature.id", "n.probes")
rpe1.data.arm.results <- rpe1.data.arm.results %>% left_join(n.probes.arm, by = "feature.id")
```

```{r}
exp4.arm.results <- readRDS(here("datasets", "exp04.rpe1.arm.accuracy.rates.RDS"))
```

```{r}
combined.arm.results <- rpe1.data.arm.results %>% inner_join(exp4.arm.results %>% select(feature.id, pct.concordance, n.probes), by = "feature.id", suffix = c(".v3", ".v2")) %>% select(!c("ploidy.concordance", "n"))
```

```{r}
combined.arm.results <- combined.arm.results %>% mutate(delta.acc = pct.concordance.v3 - pct.concordance.v2, 
                            delta.probes = n.probes.v3 - n.probes.v2) %>% mutate(feature.label = str_sub(feature.id, 4,-1))
```

```{r}
rpe1.arm.sensitivity.change <- 
ggplot(combined.arm.results, aes(x = feature.id, y = delta.acc)) +
    geom_col(fill = "#7EBDC3", color = "black", width = .75) +
    theme_bw_custom +
    theme(axis.text.x = element_text(hjust = 1, angle = 45), panel.border = element_blank(), axis.line = element_line(), panel.grid.minor.y = element_blank()) + 
    labs(x = "Chromosome Arm", y = "\u0394 Accuracy (pp)") +
    scale_y_continuous(limits = c(-10, 30), expand = c(0,0,0,0)) +
    geom_text(aes(y = 28, label = sprintf("%+2d", delta.probes)), size = 6/.pt) +
    geom_hline(yintercept = 0) +
    scale_x_discrete(labels = combined.arm.results$feature.label)
rpe1.arm.sensitivity.change
```

```{r}
saveRDS(rpe1.arm.sensitivity.change, here("figures", "figure-l", "exp10.rpe1.sensitivity.arm.change.RDS"))
```

```{r}
ggplot(combined.arm.results, aes(x = delta.probes, y = delta.acc)) +
    geom_point() + 
    geom_smooth(method = "lm", se = F) + 
    theme_bw() + 
    labs(x = "Change in number of probes", y = "Change in accuracy (pct points)") +
    scale_x_continuous(breaks = seq(-4, 8, 2)) +
    scale_y_continuous(limits = c(-15, 30))
```

```{r}
summary(lm(data = combined.arm.results, delta.acc ~ delta.probes))
cor(combined.arm.results$delta.probes, combined.arm.results$delta.acc)
```

### Theoretical Sensitivity

```{r}
exp4.gmm.metrics.arms <- readRDS(here("datasets", "exp4.gmm.metrics.arms.RDS"))
```

```{r}
exp10.gmm.metrics.arms <- calcTruePosRate(exp10, "arm") %>% filter(cn.sim.class != "sim_cn6")
exp10.gmm.metrics.arms <- exp10.gmm.metrics.arms %>% left_join(n.probes.arm, by = "feature.id")
exp10.gmm.metrics.arms <- exp10.gmm.metrics.arms %>% mutate(false.neg.p = false.neg.p * 100, true.pos.p = true.pos.p * 100)
```

```{r}
tpr.arm.bars <- 
ggplot(exp10.gmm.metrics.arms, aes(x = feature.id, y = true.pos.p)) +
    geom_col(aes(fill = cn.sim.class), color = "black", width = 0.75) +
    facet_wrap(cn.sim.class ~., ncol = 1) +
    theme_classic_custom +
    theme(legend.position = "none", panel.grid.major = element_line(), 
          strip.background = element_blank(), 
          axis.text.x = element_text(angle = 45, hjust=1)) +
    scale_fill_manual(values = c("#785EF0", "#648FFF", "#DC267F", "#FE6100", "#FFB000")) +
    labs(x = "Chromosome Arm", y = "Theoretical Sensitivity (%)") +
    scale_y_continuous(breaks = c(0, 50, 100), expand = c(0,0,0,10)) +
    scale_x_discrete(labels = rpe1.data.arm.results$feature.label) +
    geom_hline(yintercept = 0) +
    geom_text(aes(label=round(true.pos.p,0)), vjust = -1, size = 5/.pt, angle = 45, hjust = 0)
tpr.arm.bars
```

```{r}
#saveRDS(tpr.arm.bars, here("figures", "figure-l", "exp10.rpe1.tpr.arm.RDS"))
```

```{r}
tpr.arm.nprobes <- 
    ggplot(exp10.gmm.metrics.arms, aes(x=n.probes, y=true.pos.p, color = cn.sim.class)) +
    geom_point(alpha = 0.5) + 
    #geom_point(data = rpe1.data.arm.results, aes(x=n.probes, y=pct.concordance), alpha = 0.5, inherit.aes = F) + 
    theme_classic_custom +
    theme(legend.position = "top", panel.grid.major = element_line()) +
    geom_smooth(se = F, span = 1) +
    geom_smooth(data = rpe1.data.arm.results, aes(x=n.probes, y=pct.concordance), 
                color = "black", inherit.aes = F, se = F, span = 1, linetype = "dashed") +
    scale_color_manual(values = viridisLite::viridis(n=5), labels = 1:5) +
    labs(x = "Number of Probes", y = "Theoretical Sensitivity (%)", color = "Copy Number")
tpr.arm.nprobes
```

```{r}
sensitivity.compare.arms <- inner_join(exp4.gmm.metrics.arms %>% select(feature.id, cn.sim.class, true.pos.p), 
           exp10.gmm.metrics.arms %>% select(feature.id, cn.sim.class, true.pos.p),
           by = c("feature.id", "cn.sim.class"), suffix = c(".exp4", ".exp10")) %>% 
    mutate(delta.tpr = true.pos.p.exp10 - true.pos.p.exp4) %>% 
    inner_join(combined.arm.results %>% select(feature.id, delta.probes), by = "feature.id")
```

```{r}
tpr.arms.bars.compare <- 
    ggplot(sensitivity.compare.arms, aes(x = feature.id, y = delta.tpr)) +
    geom_col(aes(fill = cn.sim.class), color = "black", width = 0.85) +
    facet_wrap(cn.sim.class ~., ncol = 1) +
    theme_bw_custom +
    theme(legend.position = "none", panel.grid.major = element_line(), 
          strip.background = element_blank(), axis.text.x = element_text(angle = 45, hjust = 1)) +
    scale_fill_viridis_d() +
    labs(x = "Chromosome Arm", y = "\u0394 Theoretical Sensitivity (pp)") +
    scale_y_continuous(limits = c(-20, 30), breaks = c(-10, 0, 10, 20, 30), expand = c(0,0,0,10)) +
    geom_text(aes(y = 30, label = sprintf("%+2d", delta.probes)), size = 6/.pt)
tpr.arms.bars.compare
```

```{r}
sensitivity.compare.avg.arms <- sensitivity.compare.arms %>% group_by(cn.sim.class) %>% summarize(mean.tpr4 = mean(true.pos.p.exp4), mean.tpr10 = mean(true.pos.p.exp10))
sensitivity.compare.avg.arms$cn.sim.class.label <- paste0("Copy Number = ", 1:5)
```

```{r}
sensitivity.compare.plot.arms <- 
ggplot(sensitivity.compare.avg.arms, aes(x = mean.tpr4, y = mean.tpr10, color = cn.sim.class, label = cn.sim.class.label)) +
    geom_point() +
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "red") + 
    theme_classic_custom +
    theme(panel.grid.major = element_line()) + 
    scale_x_continuous(limits = c(20, 100), expand = c(0,0,0,0)) + 
    scale_y_continuous(limits = c(20, 100), expand = c(0,0,0,0)) +
    geom_text(nudge_x = -3, nudge_y = 2, size = 6/.pt, color = "black") +
    guides(color = "none") + 
    scale_color_manual(values = c("#785EF0", "#648FFF", "#DC267F", "#FE6100", "#FFB000")) +
    labs(x = "Panel Version 2\nMean Theoretical Sensitivity (%)", y = "Panel Version 3\nMean Theoretical Sensitivity (%)")
sensitivity.compare.plot.arms
``` 

```{r}
saveRDS(sensitivity.compare.plot.arms, here("figures", "figure-l", "exp10.rpe1.sensitivity.compare.arms.RDS"))
```

### Cross Validation

```{r}
exp10.cv <- exp10
control.copy.number <- generateControlCopyNumberTemplate(exp10, copy.number = 2, "RPE1")
control.copy.number["chr10q", "copy.number"] <- 3
```

```{r}
cv.results.arms <- map(levels(cv.folds$cv.fold), function(x){
    
    training.bcs <- cv.folds %>% filter(cv.fold != x) %>% pull(cell.barcode)
    exp10.cv <- suppressMessages(calcGMMCopyNumber(exp10.cv, cell.barcodes = training.bcs, control.copy.number = control.copy.number))
    
    rpe1.cv.data <- getTidyData(exp10.cv, "smoothedCopyNumberByArm", "gmmCopyNumber") %>% filter(cluster == "RPE1", !cell.barcode %in% training.bcs) %>% select(!c("total.reads", "cluster"))
    rpe1.cv.data$known.ploidy <- 2
    rpe1.cv.data$known.ploidy[rpe1.cv.data$feature.id == "chr10q"] <- 3
    rpe1.cv.data.results.arms <- rpe1.cv.data %>% group_by(cell.barcode, feature.id) %>% summarise(cn.concordance = gmmCopyNumber == known.ploidy, .groups = "drop") %>% group_by(feature.id) %>% dplyr::count(cn.concordance) %>% mutate(pct.concordance = round(n*100/sum(n), 1)) %>% filter(cn.concordance == TRUE) %>% add_column(test.fold = x)
    
}) %>% list_rbind() %>% left_join(n.probes.arm, by = "feature.id")
cv.results.arms$feature.label <- cv.results.arms$feature.id %>% str_sub(4,-1)
```

```{r}
cv.results.arms %>% 
    ggplot(aes(x = feature.id, y = pct.concordance)) +
    ggdist::stat_pointinterval(.width = c(1), linewidth = 2, point_size = 1, point_interval = "mean_qi") + 
    geom_point(position = position_nudge(x=-0.3), alpha = 0.5, size = 1) +
    geom_point(data = rpe1.data.arm.results, aes(x=feature.id, y=pct.concordance), inherit.aes = F, pch = 4, color = "red", position = position_nudge(x=0.3)) +
    theme_bw_custom +
    scale_y_continuous(limits = c(34, 100)) +
    labs(y = "Accuracy (%)", x = "Chromosome") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    scale_x_discrete(labels = cv.results.arms$feature.label)
```

```{r}
arm.sensitivity.mad <- 
cv.results.arms %>% group_by(feature.id, feature.label) %>% summarize(mean.pct = mean(pct.concordance), mad = mad(pct.concordance), mean.ad = DescTools::MeanAD(pct.concordance), .groups = "drop") %>% print() %>% 
    ggplot(aes(x = feature.id, y = mean.pct)) + 
    ggdist::geom_pointinterval(aes(ymin = mean.pct - mean.ad, ymax = mean.pct + mean.ad), shape = 16) +
    theme_classic_custom + 
    theme(panel.grid.major = element_line(), axis.text.x = element_text(angle = 45, hjust = 1)) +
    scale_y_continuous(limits = c(60, 100), expand = c(0,0,0,0), breaks = seq(50, 100, 10)) +
    scale_x_discrete(labels = cv.results.arms$feature.label) +
    labs(y = "Accuracy (%)", x = "Chromosome Arm") + 
    geom_hline(yintercept = mean(cv.results.arms$pct.concordance), linetype = "dashed", alpha = 0.7) +
    annotate("text", x = 40, y = 95, size = 7/.pt, label = paste0("avg: ", round(mean(cv.results.arms$pct.concordance), 0), "%"))
arm.sensitivity.mad
```

```{r}
saveRDS(arm.sensitivity.mad, here("figures", "figure-l", "exp10.rpe1.sensitivity.arm.mad.RDS"))
```

### 3 Component

```{r}
exp10.gmm.metrics.arm.3comp <- calcTruePosRate(exp10.3comp, "arm")
exp10.gmm.metrics.arm.3comp <- exp10.gmm.metrics.arm.3comp %>% left_join(n.probes.arm, by = "feature.id")
exp10.gmm.metrics.arm.3comp <- exp10.gmm.metrics.arm.3comp %>% mutate(false.neg.p = false.neg.p * 100, true.pos.p = true.pos.p * 100)
exp10.gmm.metrics.arm.3comp$feature.label <- exp10.gmm.metrics.arm.3comp$feature.id %>% str_sub(4,-1)
```

```{r}
exp10.gmm.metrics.arm.avg.3comp <- exp10.gmm.metrics.arm.3comp %>% group_by(cn.sim.class) %>% summarize(mean.tpr = mean(true.pos.p), .groups = "drop") %>% print()
```

```{r}
exp10.3comp.arm.sens <- 
ggplot(exp10.gmm.metrics.arm.3comp, aes(x = feature.id, y = true.pos.p)) +
    geom_col(aes(fill = cn.sim.class), color = "black", width = 0.75) +
    facet_wrap(cn.sim.class ~., ncol = 1) +
    theme_classic_custom +
    theme(legend.position = "none", panel.grid.major.y = element_line(),
          strip.background = element_blank(), 
          axis.text.x = element_text(angle = 45, hjust=1)) +
    scale_fill_manual(values = c("#785EF0", "#648FFF", "#DC267F")) +
    labs(x = "Chromosome Arm", y = "Theoretical Sensitivity (%)") +
    scale_y_continuous(breaks = c(0, 50, 100), expand = c(0,0,0,0)) +
    scale_x_discrete(labels = exp10.gmm.metrics.arm.3comp$feature.label) +
    geom_text(aes(label=round(true.pos.p,0)), vjust = -1, size = 6/.pt) +
    geom_text(data = exp10.gmm.metrics.arm.avg.3comp, aes(label = glue::glue("avg: {round(mean.tpr, 0)}%"), x = 38, y = 120), size = 6/.pt, inherit.aes = F)
exp10.3comp.arm.sens
```

```{r}
#saveRDS(exp10.3comp.arm.sens, here("figures", "figure-l", "exp10.rpe1.tpr.arm.3comp.RDS"))
```

```{r}
ggplot(exp10.gmm.metrics.arm.3comp, aes(x=n.probes, y=true.pos.p, color = cn.sim.class)) +
    geom_point(alpha = 0.5, position = position_jitter(width = 0.25)) + 
    theme_classic_custom +
    theme(legend.position = "top", panel.grid.major = element_line()) +
    theme(legend.box.margin = margin(0, 10, 0, 10), legend.margin = margin(0), plot.margin = unit(c(0, 5.5, 5.5, 5.5), "pt")) +
    geom_smooth(se = F, span = 1, linewidth = 0.75) +
    geom_smooth(data = rpe1.data.arm.results, aes(x=n.probes, y=pct.concordance), 
                color = "black", inherit.aes = F, se = F, span = 1, linetype = "dashed", linewidth = 0.75) +
    scale_color_manual(values = c("#785EF0", "#648FFF", "#DC267F", "#FE6100", "#FFB000"), labels = 1:5) +
    labs(x = "Number of Probes", y = "Theoretical Sensitivity (%)", color = "Copy Number")
```

```{r}
exp10.gmm.specificity.arm.3comp <- calcTrueNegRate(exp10.3comp, chromosome.scope = "arm") %>% mutate(tnr = tnr * 100)
exp10.gmm.specificity.arm.3comp <- exp10.gmm.specificity.arm.3comp %>% left_join(n.probes.arm, by = "feature.id")

ggplot(exp10.gmm.specificity.arm.3comp, aes(x = feature.id, y = tnr)) +
    geom_col(aes(fill = cn.sim.class), color = "black", width = 0.85) +
    facet_wrap(cn.sim.class ~., ncol = 1) +
    theme_classic_custom +
    theme(legend.position = "none", panel.grid.major.y = element_line(), 
          strip.background = element_blank(), axis.text.x = element_text(angle = 45, hjust = 1)) +
    scale_fill_manual(values = c("#785EF0", "#648FFF", "#DC267F")) +
    labs(x = "Chromosome", y = "Theoretical Specificity (%)") +
    scale_y_continuous(breaks = c(0, 50, 100), expand = c(0,0,0,25)) +
    scale_x_discrete(labels = exp10.gmm.metrics.arm.3comp$feature.label) +
    geom_text(aes(label=round(tnr,0)), vjust = -1, size = 6/.pt)
```

```{r}
ggplot(exp10.gmm.specificity.arm.3comp, aes(x=n.probes, y=tnr, color = cn.sim.class)) +
    geom_point(alpha = 0.5, position = position_jitter(width = 0.25)) + 
    theme_classic_custom +
    theme(legend.position = "top", panel.grid.major = element_line()) +
    theme(legend.box.margin = margin(0, 10, 0, 10), legend.margin = margin(0), plot.margin = unit(c(0, 5.5, 5.5, 5.5), "pt")) +
    geom_smooth(se = F, span = 1, linewidth = 0.75) +
    scale_color_manual(values = c("#785EF0", "#648FFF", "#DC267F"), labels = 1:3) +
    labs(x = "Number of Probes", y = "Theoretical Specificity (%)", color = "Copy Number") +
    scale_y_continuous(limits = c(70, 100), expand = c(0,0,0,1))
```

# Theo Sensitivity Table

```{r}
exp10.gmm.metrics.table <- exp10.gmm.metrics %>% mutate(feature.id = paste0("chr", feature.id)) %>% rbind(exp10.gmm.metrics.arms) %>% select(feature.id, cn.sim.class, true.pos.p) %>% mutate(true.pos.p = round(true.pos.p, 2))
exp10.gmm.metrics.table <- exp10.gmm.metrics.table %>% pivot_wider(names_from = cn.sim.class, values_from = true.pos.p)
```

```{r}
exp10.gmm.metrics.3.table <- exp10.gmm.metrics.3 %>% mutate(feature.id = paste0("chr", feature.id))
exp10.gmm.metrics.arm.3comp
```

```{r}
sensitivity.gain <- exp10.gmm.metrics.3.table %>% filter(cn.sim.class == "sim_cn3") %>% pull(true.pos.p) %>% round(.,2) %>% set_names(NULL)
sensitivity.gain <- c(sensitivity.gain, exp10.gmm.metrics.arm.3comp %>% filter(cn.sim.class == "sim_cn3") %>% pull(true.pos.p) %>% round(.,2) %>% set_names(NULL))
exp10.gmm.metrics.table$gain <- sensitivity.gain
```

```{r}
write.csv(x = exp10.gmm.metrics.table, file = here("figures", "figure-l", "exp10.theoretical.sensitivity.csv"), quote = F, row.names = F)
```

# GMM Heatmap Comparison

```{r}
exp1 <- readRDS(file = here("datasets", "exp01.tapestriexperiment.RDS"))
exp4 <- readRDS(file = here("datasets", "exp04.tapestriexperiment.RDS"))
```

```{r}
exp1.rpe1 <- colData(exp1)$cell.barcode[colData(exp1)$cluster == "RPE1"]
exp4.rpe1 <- colData(exp4)$cell.barcode[colData(exp4)$cluster == "RPE1"]
exp10.rpe1 <- colData(exp10)$cell.barcode[colData(exp10)$cluster == "RPE1"]
```

```{r}
rpe1.mat <- list_cbind(list(assay(altExp(exp1[,exp1.rpe1], "smoothedCopyNumberByArm"), "gmmCopyNumber"), 
           assay(altExp(exp4[,exp4.rpe1], "smoothedCopyNumberByArm"), "gmmCopyNumber"),
           assay(altExp(exp10[,exp10.rpe1], "smoothedCopyNumberByArm"), "gmmCopyNumber")))
rownames(rpe1.mat) <- str_split_i(rownames(rpe1.mat), "chr", 2)
```

```{r}
gmm.anno.scale <- c("1" = "#2c7bb6", "2" = "#ffffff", "3" = "#fdae61", "4" = "#d7191c", "5" = "#d7191c", "6" = "#d7191c")
```

```{r}
experiment.vector.rpe1 <- factor(c(rep("RPE1 V1", length(exp1.rpe1)), rep("RPE1 V2", length(exp4.rpe1)), rep("RPE1 V3", length(exp10.rpe1))), levels = c("RPE1 V1", "RPE1 V2", "RPE1 V3"))
```

```{r}
rpe1.hm <- Heatmap(t(rpe1.mat), cluster_columns = F, col = gmm.anno.scale, cluster_rows = T, show_row_dend = F, show_column_dend = F, show_row_names = F, column_names_side = "top", name = "Copy Number\nCall", cluster_row_slices = F,
        column_names_centered = T, column_names_rot = 45, split = experiment.vector.rpe1, border = T, use_raster = F,
        heatmap_legend_param = list(title_position = "topcenter", border = "black", labels = c("4+", "3", "2", "1"), at = 4:1))
```

```{r}
exp1.lovo <- colData(exp1)$cell.barcode[colData(exp1)$cluster == "LoVo"]
exp10.lovo <- colData(exp10)$cell.barcode[colData(exp10)$cluster == "LoVo"]
```

```{r}
lovo.mat <- list_cbind(list(assay(altExp(exp1[,exp1.lovo], "smoothedCopyNumberByArm"), "gmmCopyNumber"), 
           assay(altExp(exp10[,exp10.lovo], "smoothedCopyNumberByArm"), "gmmCopyNumber")))
rownames(lovo.mat) <- str_split_i(rownames(lovo.mat), "chr", 2)
```

```{r}
experiment.vector.lovo <- factor(c(rep("LoVo V1", length(exp1.lovo)), rep("LoVo V3", length(exp10.lovo))), levels = c("LoVo V1", "LoVo V3"))
```

```{r}
lovo.hm <- Heatmap(t(lovo.mat), cluster_columns = F, col = gmm.anno.scale, cluster_rows = T, show_row_dend = F, show_column_dend = F, show_row_names = F, column_names_side = "top", name = "Copy Number\nCall", cluster_row_slices = F,
        column_names_centered = T, column_names_rot = 45, split = experiment.vector.lovo, border = T, 
        heatmap_legend_param = list(title_position = "topcenter", border = "black", labels = c("4+", "3", "2", "1"), at = 4:1))
```

```{r}
exp1.ls <- colData(exp1)$cell.barcode[colData(exp1)$cluster == "LS513"]
exp10.ls <- colData(exp10)$cell.barcode[colData(exp10)$cluster == "LS513"]
```

```{r}
ls.mat <- list_cbind(list(assay(altExp(exp1[,exp1.ls], "smoothedCopyNumberByArm"), "gmmCopyNumber"), 
           assay(altExp(exp10[,exp10.ls], "smoothedCopyNumberByArm"), "gmmCopyNumber")))
rownames(ls.mat) <- str_split_i(rownames(ls.mat), "chr", 2)
```

```{r}
experiment.vector.ls <- factor(c(rep("LS513 V1", length(exp1.ls)), rep("LS513 V3", length(exp10.ls))), levels = c("LS513 V1", "LS513 V3"))
```

```{r}
ls.hm <- Heatmap(t(ls.mat), cluster_columns = F, col = gmm.anno.scale, cluster_rows = T, show_row_dend = F, show_column_dend = F, show_row_names = F, column_names_side = "top", name = "Copy Number\nCall", cluster_row_slices = F,
        column_names_centered = T, column_names_rot = 45, split = experiment.vector.ls, border = T, 
        heatmap_legend_param = list(title_position = "topcenter", border = "black", labels = c("4+", "3", "2", "1"), at = 4:1))
```

```{r}
saveRDS(rpe1.hm, here("figures", "figure-l", "rpe1.gmm.heatmap.comparison.RDS"))
saveRDS(lovo.hm, here("figures", "figure-l", "lovo.gmm.heatmap.comparison.RDS"))
saveRDS(ls.hm, here("figures", "figure-l", "ls.gmm.heatmap.comparison.RDS"))
```

# Gains Losses Tables

```{r}
exp1.cn.results <- getTidyData(exp1, alt.exp = "smoothedCopyNumberByArm", assay = "gmmCopyNumber") %>% mutate(gmmCopyNumber = fct_collapse(as.factor(gmmCopyNumber), "1" = "1", "2" = "2", other_level = "3+")) %>% group_by(feature.id, cluster) %>% count(gmmCopyNumber, .drop = F) %>% mutate(pct = round(100*n/sum(n),1)) %>% select(!c(n)) %>%  pivot_wider(names_from = feature.id, values_from = pct) %>% print()
```

```{r}
exp4.cn.results <- getTidyData(exp4, alt.exp = "smoothedCopyNumberByArm", assay = "gmmCopyNumber") %>% mutate(gmmCopyNumber = fct_collapse(as.factor(gmmCopyNumber), "1" = "1", "2" = "2", other_level = "3+")) %>% group_by(feature.id, cluster) %>% count(gmmCopyNumber, .drop = F) %>% mutate(pct = round(100*n/sum(n),1)) %>% select(!c(n)) %>%  pivot_wider(names_from = feature.id, values_from = pct) %>% print()
```

```{r}
exp10.cn.results <- getTidyData(exp10, alt.exp = "smoothedCopyNumberByArm", assay = "gmmCopyNumber") %>% mutate(gmmCopyNumber = fct_collapse(as.factor(gmmCopyNumber), "1" = "1", "2" = "2", other_level = "3+")) %>% group_by(feature.id, cluster) %>% count(gmmCopyNumber, .drop = F) %>% mutate(pct = round(100*n/sum(n),1)) %>% select(!c(n)) %>%  pivot_wider(names_from = feature.id, values_from = pct) %>% print()
```

```{r}
write.csv(exp1.cn.results, file = here("figures", "figure-l", "exp1.cn.results.csv"), quote = F, row.names = F)
write.csv(exp4.cn.results, file = here("figures", "figure-l", "exp4.cn.results.csv"), quote = F, row.names = F)
write.csv(exp10.cn.results, file = here("figures", "figure-l", "exp10.cn.results.csv"), quote = F, row.names = F)
```

# session_info

```{r}
sessioninfo::session_info()
```