---
title: "Figure K - Exp 5 KaryoCreate"
author: "Joey Mays"
date: '2023-07-26'
output:
  pdf_document:
    toc: yes
  html_notebook:
    code_folding: show
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
    theme: flatly
    df_print: paged
  html_document:
    code_folding: show
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
    theme: flatly
    df_print: paged
editor_options:
  chunk_output_type: inline
  markdown:
    wrap: 72
---

# Setup

```{r message=FALSE}
library(karyotapR)
library(tidyverse)
library(ggplot2)
library(here)
library(cowplot)
library(ComplexHeatmap)
```

```{r}
set.seed(20230726)
```

```{r}
here::i_am("analysis-notebooks/Figure-K_exp5_karyocreate.Rmd")
```

```{r}
dir.create(here("figures", "figure-k"), showWarnings = F, recursive = T)
fig.scalar = 1.375

theme_bw_custom <- theme_bw()
theme_bw_custom$axis.text <- element_text(size = 7)
theme_bw_custom$axis.title <- element_text(size = 8)
theme_bw_custom$legend.text <- element_text(size = 7)
theme_bw_custom$legend.title <- element_text(size = 8)
theme_bw_custom$strip.text <- element_text(size = 7)
```

# Load Data

```{r}
exp5 <- readRDS(file = here("datasets", "exp05.tapestriexperiment.RDS"))
#exp5.full <- readRDS(file = here("datasets", "exp05.full.tapestriexperiment.RDS"))
```

# Analysis

```{r}
hcec.bcs <- colData(exp5) %>% as_tibble() %>% filter(cluster == "HCEC") %>% pull(cell.barcode)
exp5.subset <- exp5[, hcec.bcs]
```

## Barcoding

```{r}
grna.lookup <- read.table(here("datasets", "exp05-grna-lookup.txt"), header = T)
```

```{r}
exp5.subset <- countBarcodedReads(exp5.subset, bam.file = here("datasets", "exp05.cells.bam"), probe = "grna", barcode.lookup = grna.lookup, max.mismatch = 2, with.indels = F)
```

```{r}
exp5.subset <- callSampleLables(exp5.subset, input.features = grna.lookup$id, output.feature = "sample.call", return.table = F, min.count.threshold = 2)
```

Calculate Gini coefficients for gRNAs to find doublets

```{r}
grna.counts <- colData(exp5.subset) %>% as_tibble() %>% column_to_rownames("cell.barcode") %>% select(starts_with("g"))
grna.counts[grna.counts <= 5] <- 0 # zero out counts less than or eq to 5

colData(exp5.subset)$grna.gini <- grna.counts %>% apply(MARGIN = 1, FUN = DescTools::Gini)
```

## Doublet Filtering

~5% doublets

```{r}
grna.gini <- colData(exp5.subset) %>% as_tibble()
summary(grna.gini$grna.gini >= 0.90)
208/(3327+208)
```

Use gini coefficient to filter out gRNA doublets. 

```{r}
to.drop <- colData(exp5.subset) %>% as_tibble() %>% filter(grna.gini < 0.9) %>% pull(cell.barcode)
exp5.subset <- exp5.subset[,!colnames(exp5.subset) %in% to.drop]
```

NA cells are hCECs without gRNA counts.

```{r}
fct_count(colData(exp5.subset)$sample.call)
unlabeled <- colData(exp5.subset) %>% as_tibble() %>% filter(is.na(sample.call)) %>% pull(cell.barcode)
exp5.subset <- exp5.subset[,!colnames(exp5.subset) %in% unlabeled]
```

## KayroCreate Analysis

```{r}
colData(exp5.subset)$sample.call <- fct_relevel(colData(exp5.subset)$sample.call, c("gNC", "g6_2", "g7_1", "g20_2", "g20_3", "g20_4", "g20_6", "g20_7"))
grna.levels.control <- levels(colData(exp5.subset)$sample.call)
grna.levels <- setdiff(grna.levels.control, "gNC")
``` 

0.3% of values are CN=4.

```{r}
cn.gmm.results.prop <- getTidyData(exp5.subset, alt.exp = "smoothedCopyNumberByChr", assay = "gmmCopyNumber") %>%
    as_tibble() %>% filter(sample.call %in% grna.levels.control)
cn.gmm.results.prop$gmmCopyNumber <- as.factor(cn.gmm.results.prop$gmmCopyNumber)
summary(cn.gmm.results.prop$gmmCopyNumber)
(240 * 100)/(sum(summary(cn.gmm.results.prop$gmmCopyNumber)))
```

Collapse copy number calls of 3 and 4 to one category.

```{r}
cn.gmm.results <- getTidyData(exp5.subset, alt.exp = "smoothedCopyNumberByChr", assay = "gmmCopyNumber") %>%
    as_tibble() %>% filter(sample.call %in% grna.levels.control)
cn.gmm.results$gmmCopyNumber <- as.factor(cn.gmm.results$gmmCopyNumber)

cn.gmm.results.collapse <- cn.gmm.results
cn.gmm.results.collapse$gmmCopyNumber <- fct_collapse(cn.gmm.results.collapse$gmmCopyNumber, "3+" = c("3", "4"))

cn.gmm.results.collapse.prop <- cn.gmm.results.collapse %>% group_by(feature.id, sample.call) %>% dplyr::count(gmmCopyNumber) %>% mutate(cn.prop = round(n*100/sum(n),1)) %>% ungroup()
cn.gmm.results.collapse.prop$sample.call <- fct_relevel(cn.gmm.results.collapse.prop$sample.call, grna.levels.control)
cn.gmm.results.collapse.prop <- cn.gmm.results.collapse.prop %>% complete(nesting(feature.id), gmmCopyNumber, sample.call, fill = list(n = 0, cn.prop = 0)) %>% print(n=100)
```

Calc p-values for Fishers exact test with multiple comparisons.
Collapse into 2 copies vs all other options (i.e., diploid vs aneulploid)

```{r}
cn.gmm.results.collapse.fisher <- cn.gmm.results.collapse
cn.gmm.results.collapse.fisher$gmmCopyNumber <- fct_collapse(cn.gmm.results.collapse.fisher$gmmCopyNumber, "aneuploid" = c("3+",  "1"))
cn.gmm.results.collapse.fisher$gmmCopyNumber <- fct_recode(cn.gmm.results.collapse.fisher$gmmCopyNumber, "diploid" = "2")
```

```{r}
calc.fisher <- function(grna){
    cn.all <- cn.gmm.results.collapse.fisher %>% select(feature.id, gmmCopyNumber, sample.call) %>% group_split(.by = feature.id, .keep = F)
    cn.fisher <- data.frame(chr = c(1:22, "X"))
    cn.fisher$p.val <- cn.all %>% map(function(x){
        x <- x %>% dplyr::select(gmmCopyNumber, sample.call) %>% filter(sample.call %in% c("gNC", grna))
        x$sample.call <- fct_drop(x$sample.call)
        x <- table(x)
        fisher.test(x)
    }) %>% map("p.value") %>% unlist()
    
    cn.fisher$p.adj <- p.adjust(cn.fisher$p.val, method = "BH")
    cn.fisher$p.adj.rounded <- round(cn.fisher$p.adj, 3)
    return(cn.fisher)
}

calc.fisher.raw <- function(grna){
    cn.all <- cn.gmm.results.collapse.fisher %>% select(feature.id, gmmCopyNumber, sample.call) %>% group_split(.by = feature.id, .keep = F)
    cn.fisher <- data.frame(chr = c(1:22, "X"))
    cn.fisher$p.val <- cn.all %>% map(function(x){
        x <- x %>% dplyr::select(gmmCopyNumber, sample.call) %>% filter(sample.call %in% c("gNC", grna))
        x$sample.call <- fct_drop(x$sample.call)
        x <- table(x)
        fisher.test(x)
    }) %>% map("p.value") %>% unlist()
    
    cn.fisher$p.val.rounded <- round(cn.fisher$p.val, 3)
    return(cn.fisher)
}

fisher.result <- map(grna.levels, ~calc.fisher(.x))
fisher.result <- fisher.result %>% list_rbind(names_to = "sample")
fisher.result$sample <- factor(fisher.result$sample, levels = unique(fisher.result$sample), labels = grna.levels)
fisher.result <- fisher.result %>% dplyr::select(sample, chr, p.adj.rounded) %>% pivot_wider(names_from = sample, values_from = p.adj.rounded) %>% print()

fisher.result.raw <- map(grna.levels, ~calc.fisher.raw(.x))
fisher.result.raw <- fisher.result.raw %>% list_rbind(names_to = "sample")
fisher.result.raw$sample <- factor(fisher.result.raw$sample, levels = unique(fisher.result.raw$sample), labels = grna.levels)
fisher.result.raw <- fisher.result.raw %>% dplyr::select(sample, chr, p.val.rounded) %>% pivot_wider(names_from = sample, values_from = p.val.rounded) %>% print()
```

```{r}
cn.gmm.results.collapse.prop %>% filter(gmmCopyNumber != "2") %>% 
    mutate(gmmCopyNumber = fct_recode(gmmCopyNumber, "loss" = "1", "gain" = "3+")) %>% 
    select(!n) %>% 
    pivot_wider(id_cols = c("sample.call", "gmmCopyNumber"), names_from = feature.id, values_from = cn.prop)

cn.gmm.results.collapse.prop %>% 
    select(!n) %>% 
    pivot_wider(id_cols = c("sample.call", "gmmCopyNumber"), names_from = feature.id, values_from = cn.prop) %>% 
write.csv(., file = here("figures", "figure-k", "exp05.kc.proportions.csv"), col.names = T, row.names = F, quote = F)
```

### Plot 

Configure p-vals for plotting

```{r}
fisher.result.figure <- pivot_longer(fisher.result, starts_with("g"), values_to = "pval", names_to = "sample.call")
colnames(fisher.result.figure)[1] <- "feature.id"
fisher.result.figure$feature.id <- as.factor(fisher.result.figure$feature.id)
fisher.result.figure$sample.call <- as.factor(fisher.result.figure$sample.call)
fisher.result.figure$pval <- as.character(fisher.result.figure$pval)
fisher.result.figure$pval.plot <- fisher.result.figure$pval
fisher.result.figure$pval.plot[fisher.result.figure$pval == 0] <- "<0.001"
fisher.result.figure$pval.plot[fisher.result.figure$pval > 0.1] <- ""
g20.idx <- intersect(which(fisher.result.figure$sample.call %in% c("g20_2", "g20_3", "g20_4", "g20_6", "g20_7")),
          which(fisher.result.figure$feature.id == 20))
fisher.result.figure$pval.plot[g20.idx] <- fisher.result.figure$pval[g20.idx]
```

```{r}
cn.gmm.results.collapse.prop.compare <- cn.gmm.results.collapse.prop %>% filter(sample.call == "gNC") %>% 
    select(!c(sample.call, n)) %>% dplyr::rename(control.cn.prop = cn.prop) %>% 
    right_join(cn.gmm.results.collapse.prop, by = c("feature.id", "gmmCopyNumber")) %>% 
    replace_na(list(control.cn.prop = 0)) %>% 
    mutate(delta.prop = cn.prop - control.cn.prop) %>% filter(gmmCopyNumber != 2, feature.id %in% c(1:17, 19:22, "X"))
```

```{r}
cn.gmm.results.collapse.prop.compare$alpha <- 1
cn.gmm.results.collapse.prop.compare$alpha <- ifelse(cn.gmm.results.collapse.prop.compare$delta.prop < 0, 0.2, 1)
```

```{r}
cn.gmm.results.collapse.prop.compare %>% filter(sample.call %in% c("gNC", "g6_2"), feature.id == 6)
cn.gmm.results.collapse.prop.compare %>% filter(sample.call %in% c("gNC", "g7_1"), feature.id == 7)
fisher.result.figure %>% filter(feature.id == 20)
cn.gmm.results.collapse.prop.compare %>% filter(sample.call %in% c("gNC", "g20_3"), feature.id == 20)
cn.gmm.results.collapse.prop.compare %>% filter(sample.call %in% c("gNC", "g20_3"), feature.id == 2)
fisher.result.figure %>% filter(feature.id == 2, sample.call == "g20_3")
```


```{r}
fig.full.wide.delta <- 
cn.gmm.results.collapse.prop.compare %>% filter(sample.call != "gNC") %>% 
    ggplot(aes(x = feature.id, y = delta.prop, fill = gmmCopyNumber)) +
    geom_col(position = "dodge", aes(alpha = I(alpha))) + 
    theme_bw_custom +
    geom_hline(yintercept = 0, linewidth = 0.25) +
    facet_grid(sample.call ~ ., switch = "y") +
    scale_fill_manual(values = c("#2E93F2", "#F43239"), labels = c("Loss", "Gain")) +
    labs(x = "", y = "\u0394 Proportion (pp)", fill = "Copy Number") +
    scale_y_continuous(limits = c(-30, 30), breaks = c(-30, 0, 30)) + 
    scale_x_discrete(position = "top") + 
    theme(strip.background = element_blank(), legend.position = "top", 
          panel.grid = element_blank(), panel.spacing.y = unit(5, "mm"),
          axis.text.x = element_text(angle = 0), 
           legend.key.size = unit(10, "pt"), 
          strip.placement = "outside", axis.ticks.x = element_blank(), strip.text.y.left = element_text(angle = 0), panel.border = element_blank()) +
    geom_text(data = fisher.result.figure %>% filter(feature.id != 18), aes(label = pval.plot, x = feature.id, y = -20), inherit.aes = F, size = 6/.pt) + 
    geom_vline(xintercept = seq(1.5, 21.5, 1), linewidth = 0.25, color = "grey") +
    geom_hline(yintercept = c(-30, 30), alpha = 0.5, linewidth = 0.25, color = "grey", linetype = "dashed") +
    guides(alpha = "none")
fig.full.wide.delta
```

```{r}
saveRDS(fig.full.wide.delta, file = here("figures", "figure-k", "exp05.kc.wide.RDS"))
```

```{r}
fisher.result.table <- pivot_longer(fisher.result, starts_with("g"), values_to = "pval", names_to = "sample.call")
colnames(fisher.result.table)[2:3] <- c("sample", "p.value.adjusted")

fisher.result.raw.table <- pivot_longer(fisher.result.raw, starts_with("g"), values_to = "pval", names_to = "sample.call")
colnames(fisher.result.raw.table)[2:3] <- c("sample", "p.value.raw")

fisher.result.table.combined <- full_join(fisher.result.table, fisher.result.raw.table, by = c("chr", "sample"))
```

```{r}
write.csv(fisher.result.table.combined, file = here("figures", "figure-k", "exp05.fisher.results.csv"), quote = F, row.names = F)
```

## LOH

get g6_2 copy number data for chr 6, add AF frequencies and probe metadata

```{r}
exp5.chr6 <- getTidyData(exp5.subset, alt.exp = "smoothedCopyNumberByChr", assay = "gmmCopyNumber") %>% filter(sample.call %in% c("g6_2"), feature.id == 6) %>% select(feature.id, cell.barcode, gmmCopyNumber, sample.call)
exp5.af <- getTidyData(exp5.subset, alt.exp = "alleleFrequency") %>% select(feature.id, cell.barcode, alleleFrequency, amplicon.id, chr)
exp5.probe.data <- rowData(exp5.subset) %>% as_tibble() %>% select(probe.id, arm)
exp5.af <- exp5.af %>% left_join(exp5.probe.data, by = c("amplicon.id" = "probe.id"))
```

Calling heterozygous variants as having an AF between 20 and 80%. 

```{r}
exp5.joint <- inner_join(exp5.af, exp5.chr6, by = c("chr" = "feature.id", "cell.barcode"))
exp5.joint %>% dplyr::select(cell.barcode, gmmCopyNumber) %>% distinct() %>% pull(gmmCopyNumber) %>% as.factor() %>% fct_count()
#exp5.joint <- exp5.joint %>% filter(gmmCopyNumber %in% 1:2, )
exp5.joint$gmmCopyNumber <- as.factor(exp5.joint$gmmCopyNumber)
het.variants <- exp5.joint %>% filter(gmmCopyNumber == 2) %>% group_by(feature.id) %>% summarize(mean.af = mean(alleleFrequency), .groups = "drop") %>% mutate(het = ifelse(mean.af > 20 & mean.af < 80, T, F)) %>% filter(het == TRUE) %>% pull(feature.id)
```

Add p and q CN calls

```{r}
exp5.joint <- getTidyData(exp5.subset, alt.exp = "smoothedCopyNumberByArm", assay = "gmmCopyNumber") %>% 
    filter(sample.call == "g6_2", feature.id %in% c("chr6p", "chr6q")) %>%
    select(feature.id, cell.barcode, gmmCopyNumber) %>% pivot_wider(id_cols = cell.barcode, names_from = feature.id, values_from = gmmCopyNumber, names_prefix = "gmm.") %>% 
    right_join(exp5.joint, by = "cell.barcode")

p.variants <- exp5.joint %>% filter(arm == "chr6p") %>% pull(feature.id) %>% unique(.)
q.variants <- exp5.joint %>% filter(arm == "chr6q") %>% pull(feature.id) %>% unique(.)
```

```{r}
exp5.joint %>% filter(feature.id %in% het.variants, gmmCopyNumber %in% 1:2) %>% mutate(rAF = abs(alleleFrequency - 50)) %>% 
    ggplot(aes(x = feature.id, y = rAF, color = gmmCopyNumber)) +
    geom_point(position = position_jitterdodge(), size = 0.5) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1), legend.position = "top") +
    labs(title = "Chr6 g6-2") + 
    labs(color = "Copy Number Call", x = "Variant", y = "abs(0.5 - AF)")
```

```{r}
#ggsave(filename = here("figures", "figure-k","chr6.BAF.pdf"), 
#       device = "pdf", height = 6, width = 11.25)
```

Plot relative allele frequency

```{r}
exp5.chr6.het.g6.nohaplo <- exp5.joint %>% filter(feature.id %in% het.variants, gmmCopyNumber %in% 1:2) %>% select(feature.id, cell.barcode, alleleFrequency) %>% 
    mutate(rAF = abs(alleleFrequency - 50)) %>% select(!c(alleleFrequency)) %>% 
    pivot_wider(names_from = feature.id, values_from = rAF) %>% column_to_rownames("cell.barcode")

#simple annotation for heatmap showing gmm
ha <- rowAnnotation(wcCN = exp5.joint %>% filter(gmmCopyNumber %in% 1:2) %>% select(cell.barcode, gmmCopyNumber) %>% distinct() %>% pull(gmmCopyNumber),
                    pCN = exp5.joint %>% filter(gmmCopyNumber %in% 1:2) %>% select(cell.barcode, gmm.chr6p) %>% distinct() %>% pull(gmm.chr6p),
                    qCN = exp5.joint %>% filter(gmmCopyNumber %in% 1:2) %>% select(cell.barcode, gmm.chr6q) %>% distinct() %>% pull(gmm.chr6q),
                    col = list(wcCN = c("1" = "#47C858", "2" = "#7b3294"), 
                               pCN = c("1" = "#FF7C1F", "2" = "#005AB5", "3" = "gray", "4" = "gray"),
                               qCN = c("1" = "#4FCFC4", "2" = "#D41159", "3" = "gray", "4" = "gray")), 
                    annotation_name_gp = gpar(fontsize = 8), 
                    annotation_legend_param = list(title_gp = gpar(fontsize = 8)))
    
arm.split <- exp5.joint %>% filter(feature.id %in% het.variants, gmmCopyNumber %in% 1:2) %>% select(feature.id, arm) %>% distinct() %>% pull(arm)

chr6.nohaplo.hm <- Heatmap(as.matrix(exp5.chr6.het.g6.nohaplo), show_row_names = F, name = "AF", cluster_columns = F, show_column_names = T, left_annotation = ha, 
        col = circlize::colorRamp2(c(50, 0), c("blue", "white")), row_km = 4, border = T, 
        show_row_dend = F, column_split = arm.split, column_names_gp = gpar(fontsize = 8),
        column_title_gp = gpar(fontsize = 8), heatmap_legend_param = list(title_gp = gpar(fontsize = 8), labels_gp = gpar(fontsize = 8), at = c(0, 25, 50), title = "Relative\nAF"))
chr6.nohaplo.hm
```

```{r}
saveRDS(chr6.nohaplo.hm, file = here("figures", "figure-k", "exp05.chr6.relativeAF.hm.RDS"))
```


Whole chromosome events (arm result matches). Whole Chromosome Mean Allele Freq (mAF).

```{r}
non.arm.events <- exp5.joint %>% filter((gmm.chr6p == 2 & gmm.chr6q == 2) | gmm.chr6p == 1 & gmm.chr6q == 1) %>% pull(cell.barcode) %>% unique(.)

het.baf.results.nonarm <- exp5.joint %>% filter(feature.id %in% het.variants, cell.barcode %in% non.arm.events, gmmCopyNumber %in% 1:2) %>% 
    mutate(rAF = abs(alleleFrequency - 50)) %>% group_by(gmmCopyNumber, cell.barcode) %>% summarize(mAF = mean(rAF)) %>% mutate(het = mAF >= 40) %>% 
    count(het) %>% mutate(pct = 100*n/sum(n)) %>% mutate(rpct = paste0(round(pct, 0), "%"))

exp5.g6.meanAF.noarms <- exp5.joint %>% filter(feature.id %in% het.variants, cell.barcode %in% non.arm.events, gmmCopyNumber %in% 1:2) %>% 
    mutate(rAF = abs(alleleFrequency - 50)) %>% group_by(gmmCopyNumber, cell.barcode) %>% summarize(mAF = mean(rAF), .groups = "drop") %>% 
    ggplot(aes(x = gmmCopyNumber, y = mAF, color = gmmCopyNumber)) +
    geom_point(position = position_jitterdodge(), size = 0.5) +
    theme_bw_custom +
    theme(legend.position = "none") +
    labs(title = "g6_2", y = "Mean Relative AF") + 
    labs(color = "Copy Number Call", x = "Copy Number Call") +
    geom_hline(yintercept = 40, linetype = 'dashed') +
    scale_color_manual(values = c("#2E93F2", "black")) +
    annotate(geom = "text", label = het.baf.results.nonarm$rpct, x = c(1.2, 1.2, 2.2, 2.2), y = c(10, 45, 10, 45)) +
    scale_y_continuous(breaks = seq(0,50,10), limits = c(0,50), expand = c(0,0,0,3))
exp5.g6.meanAF.noarms
```

```{r}
exp5.nc <- getTidyData(exp5.subset, alt.exp = "smoothedCopyNumberByChr", assay = "gmmCopyNumber") %>% filter(sample.call %in% c("gNC"), feature.id == 6) %>% select(feature.id, cell.barcode, gmmCopyNumber, sample.call)
exp5.af <- getTidyData(exp5.subset, alt.exp = "alleleFrequency") %>% select(feature.id, cell.barcode, alleleFrequency, amplicon.id, chr)
exp5.probe.data <- rowData(exp5.subset) %>% as_tibble() %>% select(probe.id, arm)
exp5.af <- exp5.af %>% left_join(exp5.probe.data, by = c("amplicon.id" = "probe.id"))

exp5.nc <- inner_join(exp5.af, exp5.nc, by = c("chr" = "feature.id", "cell.barcode"))
exp5.nc$gmmCopyNumber <- as.factor(exp5.nc$gmmCopyNumber)
```

```{r}
het.baf.results.nonarm.nc <- exp5.nc %>% filter(feature.id %in% het.variants, gmmCopyNumber %in% 1:2) %>% 
    mutate(rAF = abs(alleleFrequency - 50)) %>% group_by(gmmCopyNumber, cell.barcode) %>% summarize(mAF = mean(rAF)) %>% mutate(het = mAF >= 40) %>% 
    count(het) %>% mutate(pct = 100*n/sum(n)) %>% mutate(rpct = paste0(round(pct, 0), "%"))

exp5.gNC.meanAF.noarms <- exp5.nc %>% filter(feature.id %in% het.variants, gmmCopyNumber %in% 1:2) %>% 
    mutate(rAF = abs(alleleFrequency - 50)) %>% group_by(gmmCopyNumber, cell.barcode) %>% summarize(mAF = mean(rAF)) %>% 
    ggplot(aes(x = gmmCopyNumber, y = mAF, color = gmmCopyNumber)) +
    geom_point(position = position_jitterdodge(), size = 0.5) +
    theme_bw_custom +
    theme(legend.position = "none") +
    labs(title = "gNC", y = "Mean Relative AF") + 
    labs(color = "Copy Number Call", x = "Copy Number Call") +
    geom_hline(yintercept = 40, linetype = 'dashed') +
    scale_color_manual(values = c("#2E93F2", "black")) +
    scale_y_continuous(breaks = seq(0,50,10), limits = c(0,50), expand = c(0,0,0,3))
exp5.gNC.meanAF.noarms
```

```{r}
exp5.g6.cnnloh <- cowplot::plot_grid(plotlist = list(exp5.gNC.meanAF.noarms, exp5.g6.meanAF.noarms))
exp5.g6.cnnloh
```

```{r}
saveRDS(exp5.g6.cnnloh, file = here("figures", "figure-k", "exp05.chr6.cnnloh.RDS"))
```

# session_info

```{r}
sessioninfo::session_info()
```