---
subject: Get Allele Frequency from Public RPE1 WGS data
---

 Data from SRA Accession ERR7477340, Reijns 2022.

`sratoolkit` to get fastqs of ERR7477340

```{bash}
prefetch --max-size 150000000 ERR7477340
fasterq-dump --split-files ERR7477340
```

`bwa` to align reads to reference genome

```{bash}
bwa mem -t 10 hg19_virusref_virusref2.fa ERR7477340_1.fastq ERR7477340_2.fastq > ERR7477340.sam
```

`picard` to add read group tags

```{bash}
java -jar picard.jar AddOrReplaceReadGroups \
    I=ERR7477340.sam \
    O=ERR7477340_RG.sam \
    SORT_ORDER=coordinate \
    RGID=1 \
    RGLB=rpe1ERR \
    RGPL=illumina \
    RGPU=unit1 \
    RGSM=cellines \
    CREATE_INDEX=True
```

`gatk` to mark duplicates and sort

```{bash}
gatk MarkDuplicatesSpark \
            -I ERR7477340_RG.sam \
            -O ERR7477340_marked_dups.bam
```

`gatk` to get allele frequencies

```{bash}
gatk --java-options "-Xmx4g" HaplotypeCaller \
-I ERR7477340_marked_dups.bam \
-R hg19_virusref_virusref2.fa \
-O ERR7477340.vcf.gz
```

Get AF values using `vcfR` package

```{r}
vcf <- read.vcfR("ERR7477340.vcf.gz", verbose=F)
x <- vcfR::vcfR2tidy(vcf, info_fields = c("AF"), info_only = T)$fix
saveRDS(x, "ERR7477340.AF.RDS")
```
